wordpress-mobile/WordPress-iOS
# Recommend App: Add button with icon cell component

Refs #16995
Closes #17002

## Summary
This change introduces a new `SingleButtonTableViewCell`, which looks like a single `UIButton` 😅 . There are several approaches explored (see #17002), but I ended up going with a plain `UIImageView` and `UILabel`. I've tried hard to make plain old `UIButton` work for this, but here are the reasons why I decided not to:
- `UIButton` absorbs touch events, which prevents `didSelectRow` from being triggered. 
  - There's a workaround for this by disabling the user interaction for the `UIButton`.
- The feature that will be using this class requires an icon on the leading side of the button. `UIButton` already has this built-in, but the spacing cannot be adjusted.
  - ...But of course, there's always a [workaround](https://noahgilmore.com/blog/uibutton-padding/). This modifies the `titleEdgeInsets` and `contentEdgeInsets` with a combination of positive and negative values, but it can work. Furthermore, it needs some extra adjustments for RTL because `UIEdgeInsets` are still using `left` and `right` values.
- Finally, the multi-line label support. I've found out that there are currently two approaches for this: using a [MultilineButton](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/ViewRelated/Views/MultilineButton.swift) (which overrides intrinsic size calculation), or pin the button's title label to the button's edges. [WooCommerce uses this approach](https://github.com/woocommerce/woocommerce-ios/blob/66881a554821da1820938cb6f64ead404e44d5b9/WooCommerce/Classes/Extensions/UIButton%2BHelpers.swift#L137-L144). These solutions work, but only if we're not displaying the button's image. I've tried various things, but couldn't get it to work just right.

Anyways, I've spent enough time on this and decided to proceed with `UILabel` and `UIImageView` for now. 😅 . Here's how it looks:

Description | Screenshot
--|--
The design | ![cell_design](https://user-images.githubusercontent.com/1299411/129219369-5134af5e-1e3f-41cb-8ec3-bd824a7965d8.png)
Normal | ![normal](https://user-images.githubusercontent.com/1299411/129219365-14f7da60-f652-4d8a-9af2-80ca80b5fe47.png)
RTL language | ![rtl](https://user-images.githubusercontent.com/1299411/129219373-fc726fa9-5125-4ff6-a4f6-20a48ebd56b8.png)
Large accessibility size | ![large](https://user-images.githubusercontent.com/1299411/129219377-7c0b2574-4127-4581-9e7e-af9e970cf31c.png)

## To Test
Since the component is not hooked up anywhere, you can only test this manually by modifying the code. Here's what I did in `AboutViewController`. First, register the cell in the `setupTableView()` method.
```swift
tableView.register(SingleButtonTableViewCell.defaultNib, forCellReuseIdentifier: SingleButtonTableViewCell.defaultReuseID)
```

Second, in the `cellForRow` method, we're gonna sneakily swap out the last cell with the new one:
```Swift
// ...
let row = rows[indexPath.section][indexPath.row]

if indexPath.section == rows.count - 1,
   indexPath.row == rows[indexPath.section].count - 1,
   let buttonCell = tableView.dequeueReusableCell(withIdentifier: SingleButtonTableViewCell.defaultReuseID, for: indexPath) as? SingleButtonTableViewCell {
    buttonCell.title = "Share WordPress with a friend"
    buttonCell.iconImage = .init(systemName: "square.and.arrow.up")
    buttonCell.tintColor = .primary

    return buttonCell
}

// ...
```

Lastly, go to the About screen and verify that the button is displayed correctly.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
The changes are not reachable, since the component is still unused.

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:
